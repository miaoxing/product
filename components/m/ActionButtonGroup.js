import {ButtonGroup} from '@mxjs/m-button';
import FooterBar from '@mxjs/m-footer-bar';
import ButtonTheme from '@mxjs/m-button/ButtonTheme';
import PropTypes from 'prop-types';

/**
 * 商品操作按钮组
 */
const ActionButtonGroup = ({ret, action, onClick}) => {
  const {code, createCart, createOrder} = ret;
  const isSuc = code === 0;

  const shortMessage = createCart.shortMessage || createOrder.shortMessage;

  const showCart = createCart.code === 0 && action !== 'createOrder';
  const showOrder = createOrder.code === 0 && action !== 'createCart';

  // 更新购物车暂无逻辑，直接展示即可
  if (action === 'updateCart') {
    return (
      <ButtonTheme>
        <FooterBar.Button variant="primary" onClick={onClick.bind(this, 'updateCart')}>
          确 定
        </FooterBar.Button>
      </ButtonTheme>
    );
  }

  return (
    <ButtonTheme>
      {isSuc && (
        <ButtonGroup>
          {showCart && (<FooterBar.Button variant="secondary" onClick={onClick.bind(this, 'createCart')}>
            加入购物车
          </FooterBar.Button>)}
          {showOrder && (<FooterBar.Button variant="primary" onClick={onClick.bind(this, 'createOrder')}>
            立即购买
          </FooterBar.Button>)}
        </ButtonGroup>
      )}
      {!isSuc && (
        <FooterBar.Button variant="primary" disabled>{shortMessage}</FooterBar.Button>
      )}
    </ButtonTheme>
  );
};

ActionButtonGroup.propTypes = {
  ret: PropTypes.object.isRequired,
  action: PropTypes.string,
  onClick: PropTypes.func,
};

export default ActionButtonGroup;
