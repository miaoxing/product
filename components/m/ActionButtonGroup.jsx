import {ButtonGroup} from '@mxjs/m-button';
import FooterBar from '@mxjs/m-footer-bar';
import ButtonTheme from '@mxjs/m-button/ButtonTheme';
import PropTypes from 'prop-types';
import {Ret} from 'miaoxing';

/**
 * 商品操作按钮组
 */
const ActionButtonGroup = ({ret, action, onClick}) => {
  const {createCart, createOrder} = ret;
  const isSuc = Ret.isSuc(ret);

  const shortMessage = createCart.shortMessage || createOrder.shortMessage;

  const showCart = Ret.isSuc(createCart) && action !== 'createOrder';
  const showOrder = Ret.isSuc(createOrder) && action !== 'createCart';

  // 更新购物车暂无逻辑，直接展示即可
  if (action === 'updateCart') {
    return (
      <ButtonTheme>
        <FooterBar.Button variant="primary" onClick={onClick.bind(this, 'updateCart')}>
          确 定
        </FooterBar.Button>
      </ButtonTheme>
    );
  }

  return (
    <ButtonTheme>
      {isSuc && (
        <ButtonGroup>
          {showCart && (<FooterBar.Button variant="secondary" onClick={onClick.bind(this, 'createCart')}>
            加入购物车
          </FooterBar.Button>)}
          {showOrder && (<FooterBar.Button variant="primary" onClick={onClick.bind(this, 'createOrder')}>
            立即购买
          </FooterBar.Button>)}
        </ButtonGroup>
      )}
      {!isSuc && (
        <FooterBar.Button variant="primary" disabled>{shortMessage}</FooterBar.Button>
      )}
    </ButtonTheme>
  );
};

ActionButtonGroup.propTypes = {
  ret: PropTypes.object.isRequired,
  action: PropTypes.string,
  onClick: PropTypes.func,
};

export default ActionButtonGroup;
