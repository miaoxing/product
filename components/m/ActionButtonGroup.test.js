import ActionButtonGroup from './ActionButtonGroup';
import {fireEvent, render} from '@testing-library/react';
import {Ret} from 'miaoxing';

describe('ActionButtonGroup', () => {
  test('basic', async () => {
    const handleClick = jest.fn();
    const {container, queryByText} = render(
      <ActionButtonGroup
        ret={Ret.suc({
          createCart: Ret.suc(),
          createOrder: Ret.suc(),
        })}
        onClick={handleClick}
      />,
    );
    expect(container).toMatchSnapshot();

    const createCart = queryByText('加入购物车');
    fireEvent.click(createCart);

    const createOrder = queryByText('立即购买');
    fireEvent.click(createOrder);

    expect(handleClick).toMatchSnapshot();
  });

  test('hide create cart', () => {
    const {container} = render(
      <ActionButtonGroup
        ret={Ret.suc({
          createCart: Ret.err('err'),
          createOrder: Ret.suc(),
        })}
        onClick={() => {
        }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('hide create order', () => {
    const {container} = render(
      <ActionButtonGroup
        ret={Ret.suc({
          createCart: Ret.suc(),
          createOrder: Ret.err('err'),
        })}
        onClick={() => {
        }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('update cart', () => {
    const handleClick = jest.fn();

    const {container, queryByText} = render(
      <ActionButtonGroup
        action="updateCart"
        ret={Ret.suc({
          createCart: Ret.suc(),
          createOrder: Ret.suc(),
        })}
        onClick={handleClick}
      />,
    );
    expect(container).toMatchSnapshot();

    const button = queryByText('确 定');
    fireEvent.click(button);
    expect(handleClick).toMatchSnapshot();
  });

  test('error', () => {
    const {container} = render(
      <ActionButtonGroup
        ret={Ret.err({
          message: '商品已下架',
          createCart: Ret.err({
            message: '商品已下架',
            shortMessage: '已下架',
          }),
          createOrder: Ret.err('err'),
        })}
        onClick={() => {
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
