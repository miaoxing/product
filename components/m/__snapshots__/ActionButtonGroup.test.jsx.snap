// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButtonGroup basic 1`] = `
<div>
  <taro-view-core
    class="mx-btn-theme mx-btn-theme-pill"
  >
    <taro-view-core
      class="mx-btn-group"
    >
      <taro-button-core
        class="mx-btn mx-btn-secondary mx-footer-bar-btn"
      >
        加入购物车
      </taro-button-core>
      <taro-button-core
        class="mx-btn mx-btn-primary mx-footer-bar-btn"
      >
        立即购买
      </taro-button-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`ActionButtonGroup basic 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "createCart",
      MouseEvent {
        "isTrusted": false,
      },
    ],
    Array [
      "createOrder",
      MouseEvent {
        "isTrusted": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ActionButtonGroup error 1`] = `
<div>
  <taro-view-core
    class="mx-btn-theme mx-btn-theme-pill"
  >
    <taro-button-core
      class="mx-btn mx-btn-primary mx-footer-bar-btn"
      disabled="true"
    >
      已下架
    </taro-button-core>
  </taro-view-core>
</div>
`;

exports[`ActionButtonGroup hide create cart 1`] = `
<div>
  <taro-view-core
    class="mx-btn-theme mx-btn-theme-pill"
  >
    <taro-view-core
      class="mx-btn-group"
    >
      <taro-button-core
        class="mx-btn mx-btn-primary mx-footer-bar-btn"
      >
        立即购买
      </taro-button-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`ActionButtonGroup hide create order 1`] = `
<div>
  <taro-view-core
    class="mx-btn-theme mx-btn-theme-pill"
  >
    <taro-view-core
      class="mx-btn-group"
    >
      <taro-button-core
        class="mx-btn mx-btn-secondary mx-footer-bar-btn"
      >
        加入购物车
      </taro-button-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`ActionButtonGroup update cart 1`] = `
<div>
  <taro-view-core
    class="mx-btn-theme mx-btn-theme-pill"
  >
    <taro-button-core
      class="mx-btn mx-btn-primary mx-footer-bar-btn"
    >
      确 定
    </taro-button-core>
  </taro-view-core>
</div>
`;

exports[`ActionButtonGroup update cart 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "updateCart",
      MouseEvent {
        "isTrusted": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
